pbft/pbft.cpp

```diff
@@ -0,0 +1,363 @@
+
+    op.record_prepare(msg);
+    this->maybe_advance_operation_state(op);
+}
+
+void
+pbft::handle_commit(const pbft_msg& msg)
+{
+
+    // Commit messages are never rejected, assuming  the sanity checks passed
+    pbft_operation& op = this->find_operation(msg);
+
+    op.record_commit(msg);
+    this->maybe_advance_operation_state(op);
+}
+
+void
+pbft::broadcast(const pbft_msg& msg)
+{
+    auto json_ptr = std::make_shared<bzn::message>(this->wrap_message(msg));
+
+    for (const auto& peer : this->peers)
+    {
+        this->node->send_message(make_endpoint(peer), json_ptr);
+    }
+}
+
+void
+pbft::maybe_advance_operation_state(pbft_operation& op)
+{
+    if (op.get_state() == pbft_operation_state::prepare && op.is_prepared())
+    {
+        this->do_prepared(op);
+    }
+
+    if (op.get_state() == pbft_operation_state::commit && op.is_committed())
+    {
+        this->do_committed(op);
+    }
+}
+
+pbft_msg
+pbft::common_message_setup(const pbft_operation& op)
+{
+    pbft_msg msg;
+    msg.set_view(op.view);
+    msg.set_sequence(op.sequence);
+    msg.set_allocated_request(new pbft_request(op.request));
+
+    // Some message types don't need this, but it's cleaner to always include it
+    msg.set_sender(this->uuid);
+
+    return msg;
+}
+
+void
+pbft::do_preprepare(pbft_operation& op)
+{
+    LOG(debug) << "Doing preprepare for operation " << op.debug_string();
+
+    pbft_msg msg = this->common_message_setup(op);
+    msg.set_type(PBFT_MSG_TYPE_PREPREPARE);
+
+    this->broadcast(msg);
+}
+
+void
+pbft::do_preprepared(pbft_operation& op)
+{
+    LOG(debug) << "Entering prepare phase for operation " << op.debug_string();
+
+    pbft_msg msg = this->common_message_setup(op);
+    msg.set_type(PBFT_MSG_TYPE_PREPARE);
+
+    this->broadcast(msg);
+}
+
+void
+pbft::do_prepared(pbft_operation& op)
+{
+    LOG(debug) << "Entering commit phase for operation " << op.debug_string();
+    op.begin_commit_phase();
+
+    pbft_msg msg = this->common_message_setup(op);
+    msg.set_type(PBFT_MSG_TYPE_COMMIT);
+
+    this->broadcast(msg);
+}
+
+void
+pbft::do_committed(pbft_operation& op)
+{
+    LOG(debug) << "Operation " << op.debug_string() << " is committed-local";
+    op.end_commit_phase();
+
+    this->service->commit_request(op.sequence, op.request);
+}
+
+size_t
+pbft::outstanding_operations_count() const
+{
+    return operations.size();
+}
+
+bool
+pbft::is_primary() const
+{
+    return this->get_primary().uuid == this->uuid;
+}
+
+const peer_address_t&
+pbft::get_primary() const
+{
+    return this->peer_index[this->view % this->peer_index.size()];
+}
+
+// Find this node's record of an operation (creating a new record for it if this is the first time we've heard of it)
+pbft_operation&
+pbft::find_operation(const pbft_msg& msg)
+{
+    return this->find_operation(msg.view(), msg.sequence(), msg.request());
+}
+
+pbft_operation&
+pbft::find_operation(const uint64_t& view, const uint64_t& sequence, const pbft_request& request)
```

Comment:
I don't believe you gain anything by passing POD as a const reference over a copy.
More style I guess, but I could just do pass by value and mark the variable const in the implementation.